tools/codegen/api/python.py

```diff
@@ -304,15 +395,54 @@ def output_idx(self) -> int:
     # tools/gen_pyi.py.  If you change any logic here, please
     # check that file too.
     def signature_str(self, *, skip_outputs: bool = False) -> str:
```

Comment:
(not in this PR) as we are adding more non-trivial 'presenter' methods to the data models I wonder if it feels cleaner if we move them off the data model class.